src/cpu/core_dyn_x86.cpp

```diff
@@ -310,30 +330,7 @@ Bits CPU_Core_Dyn_X86_Run() noexcept
 			int32_t old_cycles=CPU_Cycles;
 			CPU_Cycles=1;
 
-			// If host FPU has been used by the recompiled code
-			// block, then we need to save the host FPU state in
-			// dyn_dh_fpu.state and free up the host FPU for the
-			// normal core to use.
-			if (dyn_dh_fpu.state_used) {
-				gen_dh_fpu_save();
-			}
-
-			// Copy the saved state to the normal core's emulated
-			// FPU so that it can resume from where the dynamic core
-			// left. We can't skip this step even when
```

Comment:
Here I explained why state_used cannot be used to skip this copy.  We should keep this comment in the code (maybe make it even more obvious) so that people do not try to mistakenly optimize this code.

Reply:
> Here I explained why state_used cannot be used to skip this copy. 

Indeed - and with logging, I've also come to the same conclusion, and have left in the comment. Thanks.

After some digging, we can add another dirty-tracker used to  indicate if dh needs to be copied to the emulated normal FPU.

Whenever the dynamic core uses the host FPU, it sets the `state_used` to `true`.  The only way this gets flipped back to `false` is when `gen_dh_fpu_save()` is called (existing code):

``` c++
#if defined(X86_DYNFPU_DH_ENABLED)
static void gen_dh_fpu_save(void)
#if defined (_MSC_VER)
{
	__asm {
	__asm	fnsave	dyn_dh_fpu.state
	__asm	fldcw	dyn_dh_fpu.host_cw
	}
	dyn_dh_fpu.state_used=false;
	dyn_dh_fpu.state.cw|=0x3f;
}
#else
{
	__asm__ volatile (
		"fnsave		%0			\n"
		"fldcw		%1			\n"
		:	"=m" (dyn_dh_fpu.state)
		:	"m" (dyn_dh_fpu.host_cw)
		:	"memory"
	);
	dyn_dh_fpu.state_used=false;
	dyn_dh_fpu.state.cw|=0x3f;
}
#endif
#endif

```

The PR adds `maybe_copy_host_fpu_to_dh()` that sets this new dirty tracker:

``` c++
void static void maybe_copy_host_fpu_to_dh()
{
	if (dyn_dh_fpu.state_used) {
		gen_dh_fpu_save();
		dyn_dh_fpu.state_copied_to_normal = false;
	}
}
```

Which now called both on the scope closures (just like the original class did, but named to fall in line with the other copy functions):

``` c++
struct HostFpuToDhCopier {
	~HostFpuToDhCopier()
	{
		maybe_copy_host_fpu_to_dh();
	}
};
```

as well as on entry to the normal core:

``` c++
static Bits copy_fpus_and_run_normal_core() noexcept
{
	// If host FPU has been used by the recompiled code
	// block, then we need to save the host FPU state in
	// dyn_dh_fpu.state and free up the host FPU for the
	// normal core to use.

 	// if (dyn_dh_fpu.state_used) {
	//	LOG_WARNING("FPU: copying host to dh and using normal core (normal FPU now invalid)");
	// }
	maybe_copy_host_fpu_to_dh();

	// If the dynamic core's FPU state was changed and hasn't yet been
	// copied to the normal core, then copy the dyn_dh_fpu.state to the
	// normal core's emulated FPU so that it can resume from where the
	// dynamic core left.
	maybe_copy_dh_fpu_to_normal();

	// LOG_MSG("FPU: Entering normal core (most recent normal FPU state should be valid)");
	auto ret = CPU_Core_Normal_Run();
```

And as you've no doubt guessed:

``` c++
static void maybe_copy_dh_fpu_to_normal()
{
	if (dyn_dh_fpu.state_copied_to_normal) {
		// LOG_MSG("FPU: skipping dh to normal because dh hasn't changed (normal FPU still valid)");
		return;
	}

	// Pre-requisite: host FPU should be copied to the dh FPU
	assert(!dyn_dh_fpu.state_used);

	FPU_SetTag(static_cast<uint16_t>(dyn_dh_fpu.state.tag & 0xffff));
	FPU_SetCW(static_cast<uint16_t>(dyn_dh_fpu.state.cw & 0xffff));
	FPU_SetSW(static_cast<uint16_t>(dyn_dh_fpu.state.sw & 0xffff));
	TOP = FPU_GET_TOP();
	FPU_SetPRegsFrom(dyn_dh_fpu.state.st_reg);

	dyn_dh_fpu.state_copied_to_normal = true;
	// LOG_MSG("FPU: copying dh to normal because dh has changed (normal FPU now valid)");
}
```

But what about the outbound normal-to-dh copy that still always happens, doesn't that mean the dh side is dirty again?

There are two pathways:

If the normal core /did/ changed its FPU state, then that state is written back out to the DH side - so the DH side is now fully in-sync with the normal side.

If the normal core /didn't/ change its FPU state, then that state is written back out anyway (a wasted step).

On the next round(s) in the dynamic core, so long as the FPU state isn't touched, then the normal side is still in sync and we can keep skipping.

Here are some log results:

![2023-01-27_11-31_1](https://user-images.githubusercontent.com/1557255/215216500-46596098-a5c2-4008-a89e-2e08c704b682.png)
![2023-01-27_11-31](https://user-images.githubusercontent.com/1557255/215216503-a617764a-20e0-475b-9c6e-d5f0493f0887.png)
![2023-01-27_11-29](https://user-images.githubusercontent.com/1557255/215216504-69a4670a-2851-4784-99dc-354aeb098204.png)
![2023-01-27_11-28](https://user-images.githubusercontent.com/1557255/215216506-ab001e7d-4ee5-4aca-a828-0922baf73ff1.png)
![2023-01-27_11-27](https://user-images.githubusercontent.com/1557255/215216507-b39724ff-1d46-477b-8e66-923c79bbcf44.png)